harmony_py/auth.py

```diff
@@ -0,0 +1,197 @@
+    Attempts to create an authenticated session in the following order:
+    1) If ``username`` and ``password`` are not None, create a session.
+    2) If ``username`` is specified but not ``password``, prompt user for a password and create a
+    session.
+    3) If ``netrc_file`` is True, rely on the automatic behavior of requests to find relevant
+    credentials in a .netrc file.
+    4) Attempt to read a username and password from environment variables, either from the system
+    or from a .env file to return a session.
+
+    Parameters:
+        username (str, optional): The EDL username.
+        password (str, optional): The EDL password.
+        netrc_file (bool, optional): Whether a .netrc file should be preferred for credentials.
+
+    Returns:
+        (:obj:`SessionWithHeaderRedirection`): The authenticated requests session.
+
+    :raises MissingCredentials: No credentials were specified or found.
+    """
+
+    if username and password:
+        return SessionWithHeaderRedirection(username, password)
+    elif username and not password:
+        password = getpass()
+        return SessionWithHeaderRedirection(username, password)
+    elif netrc_file:
+        return SessionWithHeaderRedirection()
+    else:
+        if cfg.EDL_USERNAME and cfg.EDL_PASSWORD:
+            return SessionWithHeaderRedirection(cfg.EDL_USERNAME, cfg.EDL_PASSWORD)
+        else:
+            raise MissingCredentials('Authentication: No credentials found.')
+
+
+def authenticate(username: Optional[str] = None, password: Optional[str] = None,
+                 netrc_file: Optional[bool] = False,
+                 verify: Optional[bool] = True) -> Session:
+    """
+    Create are requests-futures session for authenticated HTTP calls after optionally verifying
+    credentials.
+    Attempts to create an authenticated session in the following order:
+    1) If ``username`` and ``password`` are not None, create a session.
+    2) If ``username`` is specified but not ``password``, prompt user for a password and create a
+    session.
+    3) If ``netrc_file`` is True, rely on the automatic behavior of requests to find relevant
+    credentials in a .netrc file.
+    4) Attempt to read a username and password from environment variables, either from the system
+    or from a .env file to return a session.
+
+    Parameters:
+        username (str, optional): The EDL username.
+        password (str, optional): The EDL password.
+        netrc_file (bool, optional): Whether a .netrc file should be preferred for credentials.
+        verify (bool, optional): Whether EDL credentials will be verified.
+
+    Returns:
+        (:obj:`SessionWithHeaderRedirection`): The authenticated requests session.
+
+    :raises MissingCredentials: No credentials were specified or found.
+    :raises BadAuthentication: Incorrect credentials or unknown error.
+    """
+    edl_verification_url = cfg.EDL_VERIFICATION_URL
+    num_workers = int(cfg.NUM_REQUESTS_WORKERS)
```

Comment:
What is the purpose of multiple workers in the context of EDL authentication?

Reply:
Authentication, specifically a validated authentication, is tied to the act of communicating with a server (Harmony + EDL) via a requests session object. In order to better support concurrent harmony-py-initiated Harmony jobs we've tentatively settled on requests-futures and they require a thread pool. The NUM_REQUESTS_WORKERS relates to the number of threads created. Conflating authentication and the act of communicating with Harmony keeps things simple as we likely won't need to separate the two for other reasons.